New internal command Resolve-DbaComputerName that uses Resolve-DbaNetworkName to replace Resolve-NetBiosName

andreasjordan opened the pull request:
<!-- Below information IS REQUIRED with every PR -->
## Type of Change
<!-- What type of change does your code introduce -->
 - [x] Bug fix (non-breaking change, fixes #5988 )
 - [ ] New feature (non-breaking change, adds functionality, fixes #<!--issue number--> )
 - [ ] Breaking change (effects multiple commands or functionality, fixes #<!--issue number--> )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

<!-- Below this line you can erase anything that is not applicable -->
### Purpose
<!-- What is the purpose or goal of this PR? (doesn't have to be an essay) -->

### Approach
<!-- How does this change solve that purpose -->

### Commands to test
<!-- if these are the examples in the help just note it as such -->

### Screenshots
<!-- pictures say a thousand words without typing any of it -->

### Learning
<!-- Optional -->
<!--
	Include:
	 - blog post that may have assisted in writing the code
	 - blog post that were initial source
	 - special or unique approach made to solve the problem
-->

andreasjordan commented:
Although still in an early state, I would like to have general feedback here, @potatoqualitee / @wsmelton / @niphlod / @ClaudioESSilva

potatoqualitee commented:
Seems reasonable to me!

niphlod commented:
+1

andreasjordan commented:
What I don't like is that it is now not just one line but a block:

```
try {
	$resolved = Resolve-DbaNetworkName -ComputerName $Source -Credential $Credential -EnableException
	$sourceFullComputerName = $resolved.FullComputerName
} catch {
	Stop-Function -Message "Error occurred while resolving $Source" -Category ConnectionError -ErrorRecord $_ -Target $Source
	return
}
```

The core part are two lines: Using `Resolve-DbaNetworkName` to get a full object and then just putting a specific property into the final variable.

What about a new parameter to `Resolve-DbaNetworkName` to only get the needed property?
Advantage: We can configure (with a config value?) inside of `Resolve-DbaNetworkName` what property is to be returned. So I can configure to use FQDN instead of FullComputerName without changing code.

And do we need `-EnableException` and `try {} catch {}` or can `Resolve-DbaNetworkName` just silently (only verbose output to analyse problems) return the input ComputerName on failure?

These both together would reduce it back to one line like it was with Resolve-NetBiosName.

I will think about this over the weekend, but if it's the total wrong direction, please stop me...

niphlod commented:
that sounds like a job for an internal function: lots of cons (refactoring, mostly, to avoid duped code) for a smaaaall upside (If I got correctly you wanna end up with `$sourceFullComputerName= Resolve-DbaNetworkName -ComputerName $Source -Credential $Credential -EnableException -giveMejustTheFullName;` ). 
Adding a param to Resolve-Dba.. isn't going to cut it, it's a public function with a wealth of good infos.

potatoqualitee commented:
What about modifying Resolve-NetBiosName and the code inside is the block you want to avoid typing every time? 

```
try {
	$resolved = Resolve-DbaNetworkName -ComputerName $Source -Credential $Credential -EnableException
	$sourceFullComputerName = $resolved.FullComputerName
} catch {
	Stop-Function -Message "Error occurred while resolving $Source" -Category ConnectionError -ErrorRecord $_ -Target $Source
	return
}
```

andreasjordan commented:
I tried to understand the gererall concept and therefor had a look at the other commands that use `Resolve-DbaNetworkName` and `Invoke-Command2`.

These external commands use `Resolve-DbaNetworkName`:
Copy-DbaCredential
Copy-DbaInstanceAudit 
Copy-DbaLinkedServer
Disable-DbaForceNetworkEncryption
Disable-DbaHideInstance
Enable-DbaForceNetworkEncryption
Enable-DbaHideInstance 
Expand-DbaDbLogFile
Export-DbaCredential 
Export-DbaLinkedServer
Get-DbaClientProtocol
Get-DbaComputerSystem 
Get-DbaExtendedProtection
Get-DbaFilestream
Get-DbaForceNetworkEncryption 
Get-DbaHideInstance
Get-DbaInstanceProtocol 
Get-DbaLocaleSetting
Get-DbaMemoryUsage
Get-DbaNetworkActivity
Get-DbaOperatingSystem
Get-DbaPowerPlan
Get-DbaRegServer
Get-DbaService
Get-DbaStartupParameter 
Get-DbaUptime
Install-DbaInstance
Move-DbaDbFile
New-DbaComputerCertificate
New-DbaComputerCertificateSigningRequest
Remove-DbaNetworkCertificate
Rename-DbaDatabase 
Set-DbaExtendedProtection
Set-DbaNetworkCertificate
Set-DbaPowerPlan
Test-DbaConnection
Test-DbaDiskAlignment
Test-DbaDiskAllocation
Test-DbaPowerPlan
Test-DbaSpn
Update-DbaInstance
Update-DbaServiceAccount

These internal commands use `Resolve-DbaNetworkName`:
Get-DecryptedObject
Get-SqlFileStructure
Invoke-ManagedComputerCommand 
Resolve-SqlIpAddress

These external commands use `Invoke-Comnmand2` but not `Resolve-DbaNetworkName` (but they should, don't they?):
Add-DbaComputerCertificate
Add-DbaPfDataCollectorCounter
Clear-DbaConnectionPool
Get-DbaClientAlias
Get-DbaComputerCertificate
Get-DbaFeature
Get-DbaManagementObject
Get-DbaMsdtc 
Get-DbaNetworkCertificate
Get-DbaPfAvailableCounter
Get-DbaPfDataCollectorSet
Get-DbaPrivilege
Get-DbaProductKey
Import-DbaPfDataCollectorSetTemplate
Invoke-DbaDbLogShipping
New-DbaClientAlias
Remove-DbaClientAlias
Remove-DbaComputerCertificate
Remove-DbaPfDataCollectorCounter
Remove-DbaPfDataCollectorSet 
Set-DbaPrivilege
Start-DbaPfDataCollectorSet
Stop-DbaPfDataCollectorSet
Test-DbaConnection
Test-DbaManagementObject 

These internal commands use `Invoke-Command2` but not `Resolve-DbaNetworkName` (which is not a problem, because they should already get a full computer name):
Get-SQLInstanceComponent
Initialize-CredSSP
Invoke-Program
Register-RemoteSessionConfiguration
Restart-WinRMService
Test-PendingReboot
Unregister-RemoteSessionConfiguration


I thought about moving the resolve into Invoke-Command2, but there are some commands that issue more than one Invoke-Command2 to the same computer - that would result in multiple resolves. So I don't want to go that way.

But as the different commands don't follow one pattern in using `Resolve-DbaNetworkName` we should maybe introduce a replacement for `Resolve-NetBiosName` (as Chrissy suggested) and change all these commands to use the replacement.

So I would like to talk with you about what the new internal commands has to do and what it should be called.

To be most easy to use, it should take a [DbaInstanceParameter] as ComputerName and optionally a [PSCredential] as Credential, and return something that can be used as a ComputerName in Invoke-Command2. As we only need the plain ComputerName, the command should return just a string that we can use in a verbose message and pass on to Invoke-Command2.
The command should then try to run `Resolve-DbaNetworkName`, in the event of a failure `Resolve-DbaNetworkName` with `-Turbo` and in the event of a failure just return `.ComputerName` from the input. The command should only output verbose messages about the failures to not bother the user. Or should we output warnings because if the resolving fails, the following `Invoke-Command2` would maybe fail also?
The default would be to return the resolved `.FullComputerName` but I think a config parameter would be nice to change that to something else like `.FQDN`. We already have the config parameter "commands.resolve-dbanetworkname.bypass" to disable the resolving completely.

So we need an name for the command and a name for the returned string in the commands that sue the command.
How about `Resolve-DbaComputerName` and `$resolvedComputerName`?

andreasjordan commented:
Changed my mind about the config and added a parameter to return other properties than FullComputerName. 
Used the new command in some of the commands, so that you can have a look...

andreasjordan commented:
I used the input $SqlInstance when possible and not the $server, as the $server.ComputerName is only .NetName of the SMO.

potatoqualitee commented:
@niphlod and @wsmelton can you give updated feedback based on the changed requirements? Merci 🙇🏼

andreasjordan commented:
For information: New internal command is Resolve-DbaComputerName, I used it in Copy-DbaCredential, Copy-DbaInstanceAudit and Copy-DbaLinkedServer - so you can have a look how I think it should be used.

andreasjordan commented:
We should include the rule "do not add domain part for a local host" that helped with Install-DbaInstance (https://github.com/sqlcollaborative/dbatools/issues/6976#issuecomment-880562827 / #7564).

andreasjordan commented:
Is this the right way? Then I would change the other commands touched here and open the pull request for review.

potatoqualitee commented:
i like it. should we keep it until after 1.1.0 or include it?

andreasjordan commented:
Does not have to be in 1.1.0 - just want to get this finished in the next days. Will now continue to work on this.

potatoqualitee commented:
@niphlod  or @wsmelton can we get your detail oriented eyes to approve this PR?

niphlod commented on functions/Copy-DbaLogin.ps1:
```diff
@@ -222,7 +222,8 @@ function Copy-DbaLogin {
             }
 
-            $serverName = Resolve-NetBiosName $sourceServer
+            # Here we don't need the FullComputerName, but only the maschine name to compare to the host part of the login name. So ComputerName should be fine.
```
typo, machine instead of maschine

andreasjordan replied:
Thanks. Fixed.

niphlod commented on functions/Sync-DbaLoginPermission.ps1:
```diff
@@ -106,7 +106,8 @@ function Sync-DbaLoginPermission {
                 }
 
-                $serverName = Resolve-NetBiosName $sourceServer
+                # Here we don't need the FullComputerName, but only the maschine name to compare to the host part of the login name. So ComputerName should be fine.
```
maschine --> machine

andreasjordan replied:
Thanks. Fixed.

niphlod reviewed (changes requested):
apart from typos, seems fine.
I'm still not a fan of to Resolve-DbaComputerName and the -Property param but till it's internal there's no reason to be nitpicky.

potatoqualitee commented:
Thank you 🙇🏼

potatoqualitee reviewed (approved):
approved per niph

potatoqualitee commented:
Here we goooo!
